File: packages/renderer-web/src/transition.ts
/**
 * The transition engine — the browser half of F7's interpolation pass (decision
 * 2026-08-11, ZAB-33). Pure and clock-injected: the view hands it a node's freshly
 * resolved target values plus `now`, and gets back the values to render this frame.
 *
 * It interpolates DECLARED INPUTS, never computed rects: the view writes the result
 * onto the node and then runs its normal measure/arrange pass with those numbers, so
 * there is one layout pass per frame and the layout never feeds back into its own
 * input. Curves come from `easeProgress` in `@zabloo/format` (the normative reference
 * implementation) — that shared arithmetic is what keeps web and Unity on the same
 * number every frame.
 */

import { type Easing, easeProgress } from "@zabloo/format";
import type { Color } from "./tessellator.js";

/** Animatable colors — lerped componentwise in straight sRGB with straight alpha. */
export type ColorProp = "background" | "borderColor" | "color";
/** Animatable scalars: paint values plus the layout dims, all lerped after token resolution. */
export type ScalarProp =
  | "opacity"
  | "radius"
  | "borderWidth"
  | "width"
  | "height"
  | "gap"
  | "padding";
export type AnimatableProp = ColorProp | ScalarProp;
export type AnimValue = number | Color;

/**
 * Scalars a COMPONENT'S BEHAVIOR tweens with endpoints it computes itself, rather
 * than values of the animatable set (decision 2026-08-11 §5): `progress` is the
 * ProgressBar's fraction, which is why the bar interpolates its value and never its
 * fill rect. They ride the same tracks as the declared props — one interruption rule,
 * one clock, one set of curves — under keys the animatable set cannot collide with.
 */
export type BehaviorKey = "progress";
export type TrackKey = AnimatableProp | BehaviorKey;

/** Iteration order of the animatable set — the normative list, nothing else animates. */
export const ANIMATABLE_PROPS: readonly AnimatableProp[] = [
  "background",
  "borderColor",
  "color",
  "opacity",
  "radius",
  "borderWidth",
  "width",
  "height",
  "gap",
  "padding",
];

/**
 * A node's animatable values for one frame: the targets the view resolves, and the
 * interpolated result `stepNode` hands back. Everything else (`fontSize`, `grow`, the
 * layout enums, every structural prop) snaps and never passes through here.
 */
export interface ResolvedValues {
  /** Absent = not declared: nothing is painted, so there is no honest endpoint — it snaps. */
  background?: Color;
  borderColor?: Color;
  color?: Color;
  /**
   * The renderer has a default for these (opacity 1, the rest 0), so both endpoints
   * always resolve to a number and they never snap for lack of a declared value.
   */
  opacity?: number;
  radius?: number;
  borderWidth?: number;
  gap?: number;
  padding?: number;
  /** Absent = auto (the measure pass decides): an auto endpoint snaps (decision §3). */
  width?: number;
  height?: number;
}

/** A tween in flight for one property. */
interface Track {
  from: AnimValue;
  to: AnimValue;
  startedAt: number;
  duration: number;
  easing: Easing;
}

/**
 * Per-node animation state. It lives on the layout node, so rebuilding the tree —
 * an envelope reload — drops it and everything snaps, which is exactly the rule:
 * transitions live INSIDE the life of one loaded document.
 */
export interface NodeAnim {
  tracks: Map<TrackKey, Track>;
  /** Last value handed to the renderer, per key — what an interruption tweens from. */
  current: Map<TrackKey, AnimValue>;
}

/** A node's `transition` with its `Dim` duration already resolved to milliseconds. */
export interface ResolvedTransition {
  duration: number;
  easing: Easing;
}

export function createNodeAnim(): NodeAnim {
  return { tracks: new Map(), current: new Map() };
}

/** Forgets everything: the next step snaps, like a mount. */
export function clearNodeAnim(anim: NodeAnim): void {
  anim.tracks.clear();
  anim.current.clear();
}

/**
 * Advances one node by one frame and returns the values to render with.
 *
 * A tween starts whenever a resolved value CHANGES, whatever caused the change —
 * there is no trigger list. It snaps when either endpoint is `undefined`, on the
 * first step (mount: no previous value), and when the node declares no usable
 * `transition`. An interruption retargets from the value on screen over a FULL
 * duration (the CSS model), so releasing a button mid-press leaves from the color
 * actually visible instead of snapping back or exiting unnaturally fast.
 */
export function stepNode(
  anim: NodeAnim,
  targets: ResolvedValues,
  transition: ResolvedTransition | null,
  now: number,
): { values: ResolvedValues; animating: boolean } {
  const values: ResolvedValues = {};
  let animating = false;

  for (const prop of ANIMATABLE_PROPS) {
    const stepped = stepTrack(anim, prop, targets[prop], transition, now);
    if (stepped.value === undefined) continue;
    animating ||= stepped.animating;
    // One cast for the whole loop: the prop union and the value union are correlated
    // by construction (a ColorProp only ever carries a Color), but TS cannot see it.
    (values as Record<AnimatableProp, AnimValue>)[prop] = stepped.value;
  }

  return { values, animating };
}

/**
 * The same step for a scalar a component's behavior owns (the ProgressBar's
 * fraction). It takes the endpoints already computed, so the behavior decides WHAT
 * moves while this file keeps deciding HOW — one interruption rule and one curve set
 * for declared props and behavior-driven ones alike.
 */
export function stepValue(
  anim: NodeAnim,
  key: BehaviorKey,
  target: number,
  transition: ResolvedTransition | null,
  now: number,
): { value: number; animating: boolean } {
  const stepped = stepTrack(anim, key, target, transition, now);
  return {
    value: typeof stepped.value === "number" ? stepped.value : target,
    animating: stepped.animating,
  };
}

/** Advances one track by one frame. `undefined` in ⇒ the key snaps and is forgotten. */
function stepTrack(
  anim: NodeAnim,
  key: TrackKey,
  target: AnimValue | undefined,
  transition: ResolvedTransition | null,
  now: number,
): { value: AnimValue | undefined; animating: boolean } {
  // Sample the tween in flight FIRST, so `current` is the value on screen right
  // now — that is the point a retarget below has to leave from.
  const track = anim.tracks.get(key);
  if (track) {
    const progress = (now - track.startedAt) / track.duration;
    anim.current.set(key, lerp(track.from, track.to, easeProgress(track.easing, progress)));
    if (progress >= 1) anim.tracks.delete(key);
  }

  if (target === undefined) {
    // Not declared (or auto): no endpoint to tween towards, and nothing to paint.
    anim.tracks.delete(key);
    anim.current.delete(key);
    return { value: undefined, animating: false };
  }

  const live = anim.tracks.get(key);
  const current = anim.current.get(key);
  let value: AnimValue;
  let animating = false;

  if (current === undefined) {
    value = target; // mount, or a node coming back into layout
  } else if (live) {
    if (sameValue(live.to, target)) {
      value = current; // already heading there
      animating = true;
    } else {
      value = retarget(anim, key, current, target, transition, now);
      animating = anim.tracks.has(key);
    }
  } else if (sameValue(current, target)) {
    value = current; // settled
  } else {
    value = retarget(anim, key, current, target, transition, now);
    animating = anim.tracks.has(key);
  }

  anim.current.set(key, value);
  return { value, animating };
}

/**
 * Points the property at a new target from where it currently is. Without a usable
 * transition (absent, `duration <= 0`, not finite) the change is instant — the exact
 * pre-F7 behavior.
 */
function retarget(
  anim: NodeAnim,
  key: TrackKey,
  current: AnimValue,
  target: AnimValue,
  transition: ResolvedTransition | null,
  now: number,
): AnimValue {
  if (!transition || !(transition.duration > 0) || !Number.isFinite(transition.duration)) {
    anim.tracks.delete(key);
    return target;
  }
  anim.tracks.set(key, {
    from: current,
    to: target,
    startedAt: now,
    duration: transition.duration,
    easing: transition.easing,
  });
  return current;
}

/**
 * Repeating 0..1 phase for behavior-driven loops (the Spinner's spin). The other
 * half of the machinery is the view's frame loop, which keeps scheduling while
 * anything is animating; a behavior samples this to drive its own endpoints.
 */
export function loopPhase(startedAt: number, now: number, period: number): number {
  if (!(period > 0) || !Number.isFinite(period)) return 0;
  const elapsed = now - startedAt;
  if (!(elapsed > 0)) return 0; // also catches NaN
  return (elapsed % period) / period;
}

/** Componentwise lerp in STRAIGHT sRGB with STRAIGHT alpha — no premultiply, no gamma. */
export function lerpColor(from: Color, to: Color, t: number): Color {
  return [
    from[0] + (to[0] - from[0]) * t,
    from[1] + (to[1] - from[1]) * t,
    from[2] + (to[2] - from[2]) * t,
    from[3] + (to[3] - from[3]) * t,
  ];
}

function lerp(from: AnimValue, to: AnimValue, t: number): AnimValue {
  if (typeof from === "number") return typeof to === "number" ? from + (to - from) * t : to;
  return typeof to === "number" ? to : lerpColor(from, to, t);
}

function sameValue(a: AnimValue, b: AnimValue): boolean {
  if (typeof a === "number" || typeof b === "number") return a === b;
  return a[0] === b[0] && a[1] === b[1] && a[2] === b[2] && a[3] === b[3];
}

File: core/src/transition.cpp
#include <cmath>
#include <cstddef>
#include <optional>

#include "easing.h"
#include "layout.h"
#include "transition.h"

namespace zabloo {

namespace {

size_t index_of(TrackKey key) { return static_cast<size_t>(key); }

Track *live_track(NodeAnim *anim, TrackKey key) {
  const size_t i = index_of(key);
  return anim != nullptr && anim->in_flight[i] ? &anim->tracks[i] : nullptr;
}

const AnimValue *current_of(const NodeAnim *anim, TrackKey key) {
  const size_t i = index_of(key);
  return anim != nullptr && anim->has_current[i] ? &anim->current[i] : nullptr;
}

void forget_track(NodeAnim *anim, TrackKey key) {
  if (anim != nullptr) anim->in_flight[index_of(key)] = false;
}

void set_current(NodeAnim *anim, TrackKey key, const AnimValue &value) {
  if (anim == nullptr) return;
  const size_t i = index_of(key);
  anim->current[i] = value;
  anim->has_current[i] = true;
}

AnimValue lerp(const AnimValue &from, const AnimValue &to, double t) {
  // A key carries the same kind for the life of a document, so the mismatch below
  // is unreachable — but "interpolating" a number towards a color has exactly one
  // honest answer, and it is the target.
  if (from.kind != to.kind) return to;
  if (from.kind == AnimValue::Kind::Number) {
    return AnimValue::of_number(from.number + (to.number - from.number) * t);
  }
  return AnimValue::of_color(lerp_color(from.color, to.color, t));
}

/**
 * Points a key at a new target from where it currently is. Without a usable
 * transition the change is instant — the exact pre-F7 behavior, and the path a
 * node with no `transition` at all takes for every one of its values.
 */
AnimValue retarget(NodeAnim *anim, TrackKey key, const AnimValue &current, const AnimValue &target,
                   const ResolvedTransition *transition, double now) {
  if (anim == nullptr || transition == nullptr || !transition->usable()) {
    forget_track(anim, key);
    return target;
  }
  const size_t i = index_of(key);
  anim->tracks[i] = Track{current, target, now, transition->duration, transition->easing};
  anim->in_flight[i] = true;
  return current;
}

/**
 * Advances one key by one frame. A null `target` means the value is not declared
 * this frame (an `auto` size, an undeclared color): the key snaps and is forgotten,
 * so the day it comes back is a mount rather than a resume.
 *
 * Returns whether there is a value to render; `out` carries it.
 */
bool step_track(NodeAnim *anim, TrackKey key, const AnimValue *target,
                const ResolvedTransition *transition, double now, AnimValue &out) {
  if (anim == nullptr) {
    if (target == nullptr) return false;
    out = *target;
    return true;
  }

  // Sample the tween in flight FIRST, so `current` is the value on screen right
  // now — that is the point a retarget below has to leave from.
  if (const Track *track = live_track(anim, key); track != nullptr) {
    const double progress = (now - track->started_at) / track->duration;
    const AnimValue at = lerp(track->from, track->to, ease_progress(track->easing, progress));
    set_current(anim, key, at);
    if (progress >= 1.0) forget_track(anim, key);
  }

  if (target == nullptr) {
    forget_track(anim, key);
    anim->has_current[index_of(key)] = false;
    return false;
  }

  const Track *flight = live_track(anim, key);
  const AnimValue *current = current_of(anim, key);
  AnimValue value;
  if (current == nullptr) {
    value = *target;  // mount, or a node coming back into layout
  } else if (flight != nullptr) {
    // Already heading there, or retargeted from the value on screen.
    value =
        flight->to == *target ? *current : retarget(anim, key, *current, *target, transition, now);
  } else {
    value = *current == *target ? *current  // settled
                                : retarget(anim, key, *current, *target, transition, now);
  }

  set_current(anim, key, value);
  out = value;
  return true;
}

/** Whether the key is STILL in flight after its step — the "animating" signal. */
bool in_flight(NodeAnim *anim, TrackKey key) { return live_track(anim, key) != nullptr; }

/** One always-declared scalar: the paint values and the two layout spacings. */
double step_scalar(NodeAnim *anim, TrackKey key, double target,
                   const ResolvedTransition *transition, double now, bool &animating) {
  const AnimValue in = AnimValue::of_number(target);
  AnimValue out;
  step_track(anim, key, &in, transition, now, out);
  if (in_flight(anim, key)) animating = true;
  return out.number;
}

/** An `auto` size: absent is a value the layout pass acts on, not a missing one. */
std::optional<double> step_dim(NodeAnim *anim, TrackKey key, const std::optional<double> &target,
                               const ResolvedTransition *transition, double now, bool &animating) {
  const AnimValue in = AnimValue::of_number(target.value_or(0.0));
  AnimValue out;
  const bool has = step_track(anim, key, target.has_value() ? &in : nullptr, transition, now, out);
  if (in_flight(anim, key)) animating = true;
  return has ? std::optional<double>(out.number) : std::nullopt;
}

/** An undeclared color: nothing is painted, which is not the same as painting black. */
std::optional<Color> step_color(NodeAnim *anim, TrackKey key, const std::optional<Color> &target,
                                const ResolvedTransition *transition, double now, bool &animating) {
  const AnimValue in = AnimValue::of_color(target.value_or(Color{}));
  AnimValue out;
  const bool has = step_track(anim, key, target.has_value() ? &in : nullptr, transition, now, out);
  if (in_flight(anim, key)) animating = true;
  return has ? std::optional<Color>(out.color) : std::nullopt;
}

}  // namespace

bool ResolvedTransition::usable() const { return duration > 0.0 && std::isfinite(duration); }

void clear_node_anim(NodeAnim &anim) {
  anim.in_flight.fill(false);
  anim.has_current.fill(false);
}

bool step_node(NodeAnim *anim, const ResolvedValues &targets, const ResolvedTransition *transition,
               double now, ResolvedValues &out) {
  bool animating = false;
  // Stepped into locals and assigned at the end, because `out` may BE the node's
  // own `resolved` — the reuse that keeps a steady-state frame allocation-free —
  // while `targets` is the scratch the caller filled from it moments ago.
  const std::optional<Color> background =
      step_color(anim, TrackKey::Background, targets.background, transition, now, animating);
  const std::optional<Color> border_color =
      step_color(anim, TrackKey::BorderColor, targets.border_color, transition, now, animating);
  const std::optional<Color> color =
      step_color(anim, TrackKey::Color, targets.color, transition, now, animating);
  const double opacity =
      step_scalar(anim, TrackKey::Opacity, targets.opacity, transition, now, animating);
  const double radius =
      step_scalar(anim, TrackKey::Radius, targets.radius, transition, now, animating);
  const double border_width =
      step_scalar(anim, TrackKey::BorderWidth, targets.border_width, transition, now, animating);
  const std::optional<double> width =
      step_dim(anim, TrackKey::Width, targets.width, transition, now, animating);
  const std::optional<double> height =
      step_dim(anim, TrackKey::Height, targets.height, transition, now, animating);
  const double gap = step_scalar(anim, TrackKey::Gap, targets.gap, transition, now, animating);
  const double padding =
      step_scalar(anim, TrackKey::Padding, targets.padding, transition, now, animating);

  out.background = background;
  out.border_color = border_color;
  out.color = color;
  out.opacity = opacity;
  out.radius = radius;
  out.border_width = border_width;
  out.width = width;
  out.height = height;
  out.gap = gap;
  out.padding = padding;
  return animating;
}

SteppedValue step_value(NodeAnim *anim, TrackKey key, double target,
                        const ResolvedTransition *transition, double now) {
  SteppedValue out;
  out.value = step_scalar(anim, key, target, transition, now, out.animating);
  return out;
}

double loop_phase(double started_at, double now, double period) {
  if (!(period > 0.0) || !std::isfinite(period)) return 0.0;
  const double elapsed = now - started_at;
  if (!(elapsed > 0.0)) return 0.0;  // also catches NaN
  return std::fmod(elapsed, period) / period;
}

Color lerp_color(const Color &from, const Color &to, double t) {
  // Widened to double for the arithmetic and narrowed back on store: the reference
  // has one numeric type, so mixing in `float` would round differently and drift a
  // channel away from the number the corpus recorded.
  const auto mix = [t](float a, float b) {
    const double lo = static_cast<double>(a);
    return static_cast<float>(lo + (static_cast<double>(b) - lo) * t);
  };
  return Color{mix(from.r, to.r), mix(from.g, to.g), mix(from.b, to.b), mix(from.a, to.a)};
}

}  // namespace zabloo
